feat: Add dynamic content_length support for encrypted streams

gerasim13 opened the pull request:
This PR enables dynamic content length calculation for streams where the final size is not known upfront, which is essential for supporting encrypted content (e.g., AES-128 encrypted HLS streams).

Changes:
* Added runtime updates of content_length from the stream in two key locations:
   * During the main download loop
   * In the finish_or_find_next_gap method

Context:
When working with encrypted content, the actual content length may only become known progressively as segments are decrypted. This change allows the content length to be updated dynamically from the stream's content_length() method rather than being fixed at initialization.

Closes #236

gerasim13 commented:
Here is an example program that decodes an encrypted stream: https://github.com/gerasim13/hls_client_rs/blob/main/examples/encrypted.rs

Commit:
refactor: replace Option<u64> with ContentLength enum

Introduce ContentLength enum (Static, Dynamic, Unknown) for better
representation of stream content length scenarios across all storage
providers and stream implementations.

gerasim13 committed:
refactor: enhance ContentLength with Dynamic variant and size tracking

Add DynamicLength struct to ContentLength enum to track reported size
for dynamic streams. Implement Into<u64>, PartialEq<u64>, and
PartialOrd<u64> traits for better ergonomics. Update storage providers
and examples to handle the new Dynamic variant.

gerasim13 committed:
refactor: update dynamic content length during streaming

Update ContentLength::Dynamic to reflect actual fetched data during download.
Remove unused From<u64> implementation for DynamicLength.

gerasim13 committed:
refactor: replace Into<Option<u64>> with explicit ContentLength matching

Replace implicit Into<Option<u64>> conversions with explicit pattern
matching over ContentLength variants (Static, Dynamic, Unknown) for
better clarity in HTTP, storage, and seek logic.

aschey commented:
General approach here looks good to me! Sorry for the CI issues. I think the issue is that the tests create a lot of temporary files that push it over the disk space threshold if the resolved cache is large enough. Not sure why it hasn't been an issue until now, but I can deal with it later.

gerasim13 committed:
refactor: improve ContentLength handling with new constructors and current_value method

- Add new_static, new_dynamic, and new_unknown constructors for ContentLength
- Introduce current_value() method to replace manual match expressions
- Update Dynamic length calculation to use gathered.unwrap_or(reported)
- Simplify PartialEq and PartialOrd implementations for Dynamic variant
- Replace direct ContentLength construction with constructors throughout codebase

aschey commented on crates/stream-download/src/storage/mod.rs:
```diff
@@ -9,6 +9,142 @@ pub mod memory;
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct DynamicLength {
+    /// The length of the content as reported by the server, for example, via Content-Range header.
+    pub reported: u64,
```
I don't think we want to allow consumers to modify these values. These should probably be kept private.

aschey commented on crates/stream-download/src/storage/mod.rs:
```diff
@@ -9,6 +9,142 @@ pub mod memory;
+/// content whose final size is determined only after processing,
+/// and content where the length is unknown.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum ContentLength {
```
This feels like it probably belongs in the `source` module rather than `storage`.

gerasim13 committed:
refactor: move ContentLength to source module and simplify storage API

- Move ContentLength and DynamicLength from storage to source module
- Simplify StorageProvider::into_reader_writer to accept Option<u64>
- Update StreamDownload::content_length to return Option<u64>
- Add helper constructors for ContentLength variants
- Implement conversions and comparisons between ContentLength and u64
- Update all storage implementations to use simplified signature

aschey commented:
Closing due to inactivity, feel free to reopen if there's further interest in this. Thanks!
